compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/handlers/RangeToHandler.kt

```diff
@@ -67,7 +64,23 @@ internal class RangeToHandler(private val context: CommonBackendContext) :
             if (preferJavaLikeCounterLoop || this.constLongValue == -1L) return null
         }
 
-        val irConst = this as? IrConst<*> ?: return null
+        return when (this) {
+            is IrConst<*> -> convertIrConst(this)
+            is IrCall -> convertIrCall(this)
+            else -> null
+        }
+    }
+
+    private fun convertIrCall(irCall: IrCall): IrExpression? {
+        return if (irCall.origin == IrStatementOrigin.MINUS && (irCall.getValueArgument(0) as? IrConst<*>)?.value == 1)
```

Comment:
That might be a bit too optimistic because of possible underflow.
Please, add the following test:
```
fun box(): String {
    val M1 = Int.MAX_VALUE - 2
    val M2 = Int.MIN_VALUE
    var t = 0
    for (x in M1..M2-1) {
        ++t
        if (t > 3) return "Failed: too many iterations"
    }
    return "OK"
}
```
We are not going to do proper value range analysis here. However, there are some relatively frequent practical cases when we know for sure that `expr - 1` will never underflow. In particular KT-22334 listed one such example (`array.size - 1`; can never underflow because `array.size` is always non-negative). Same is true about `collection.size` and `charSequence.size`.